presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -66,9 +65,6 @@
     private final ScheduledExecutorService executor;
 
     @GuardedBy("this")
-    private final Set<URI> locations = new HashSet<>();
-
-    @GuardedBy("this")
     private boolean noMoreLocations;
 
     private final ConcurrentMap<URI, HttpPageBufferClient> allClients = new ConcurrentHashMap<>();
```

Comment:
It looks like we only actually care about its size out of synchronized blocks. As a result, it would probably be simpler to make this object guarded and add a counter.